clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp

```diff
@@ -0,0 +1,27 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -fsyntax-only -sycl-std=2020 -verify %s
+
+// This test warns users when an ODR-use of a constexpr variable causes the kernel lambda to capture it as a
+// kernel argument
+
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+
+queue q;
+
+class LambdaKernel;
+
+int main() {
+
+  constexpr unsigned OdrUsedVar = 10;
+
+  q.submit([&](handler &h) {
+    // expected-note@+1 {{in instantiation of function template specialization}}
+    h.single_task<LambdaKernel>(
+        [=]() {
+          // constexpr 'OdrUsedVar' is odr-used here.
+          const unsigned *ptr = &OdrUsedVar; // expected-warning {{captured constexpr 'OdrUsedVar' will be a kernel argument in device code}}
```

Comment:
I don't quite get the purpose of this warning.

I think in order to understand this warning message, user must know DPC++ implementation details about [lowering lambda object into kernel function](https://intel.github.io/llvm-docs/CompilerAndRuntimeDesign.html#lowering-of-lambda-function-objects-and-named-function-objects) and how kernel function parameters are composed. Do you think it's clear for DPC++ users?

Another point it's generic C++ behavior and not SYCL specific issue:

```c++
template <typename T>
void foo(T t) {t();}
// ...
foo([=]() { const unsigned *ptr = &OdrUsedVar; }); // OdrUsedVar is captured and might cause performance issues
foo([=]() { unsigned x = 10 + OdrUsedVar; }); // OdrUsedVar is not captured. No performance issues?
```
Should we emit similar warning for any lambda expression?

It's probably more useful to implement such diagnostics in performance profiler because passing additional argument might be not a problem at all.

@premanandrao, @srividya-sundaram, @erichkeane, what do you think?

Reply:
tagging @kbobrovs to give his feedback since this was originally brought up by him.

Reply:
One more thing: this diagnostics warns that current implementation is not optimal.
I assume if compiler detects odr-use of `constexpr` it can allocate a global constant/variable in device code and kernel function object can capture global object and avoid passing it via kernel argument. This seems to be valid implementation. Right?

Reply:
> It's probably more useful to implement such diagnostics in performance profiler because passing additional argument might be not a problem at all.

I suspect many people (like myself) writing code similar to
```
  constexpr int V = 137;

  auto f = [=](Id i) -> int {
    return (i * V).x;
  };
```
will expect V to be compile-time constant inside the lambda. In some cases this expectation is not true and V is captured and hence is not a compile-time constant. This can potentially lead to performance problems.

I'm not sure what 'performance compiler' is. If it can be built from this github repo - I'm OK with making this feature available only in 'performance compiler'.

Reply:
FWIW, that is a fairly well known property of constexpr: It isn't necessarily a compile time constant unless you make it so in some way.  I'm generally not in favor of a warning that tells you that the language is still working the way it is designed, particularly since there isn't a code-way to suppress this warning.

If we had some sort of perf-analysis tool, that would be a good place for that.  Have we considered seeing if the Static Analysis effort would be a better place for this?